### PR TITLE
New Processing Algorithm: Equidistance Buffer

### PR DESCRIPTION
**Pull request by spatialthoughts**
This is a new algorithm that approximate geodesic buffers in QGIS by transforming each feature geometry to a custom Azimuthal Equidistance projection centered at that geometry, and then transforming it back to the original projection. I have written a post explaining the rationale behind this algorithm https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

### CODE REVIEW

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                       QgsProcessingException,
+                       QgsCoordinateReferenceSystem,
+                       QgsCoordinateTransform,
+                       QgsProject
+                       )
+    
+from processing.algs.qgis.QgisAlgorithm import QgisFeatureBasedAlgorithm
+
+
+class EquidistanceBuffer(QgisFeatureBasedAlgorithm):
```
Comment (elpaso) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Cool, thanks! 
May I suggest that you copy the alg description in the class docstring?

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        self.distance = None
+        self.segments = None
+        self.join_style = None
+        self.side = None
+        self.miter_limit = None
+        self.join_styles = [self.tr('Round'),
+                            'Miter',
+                            'Bevel']
+        self.end_cap_styles = [self.tr('Round'),
+                       'Flat',
```
Comment (luipir) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: why translate only 'Round' ?

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 1 above)
Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Thanks. I have added the docstring

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                       'Square']
+
+    def initParameters(self, config=None):
+        self.addParameter(QgsProcessingParameterDistance(self.DISTANCE,
+                                                         self.tr('Distance (in meters'),
```
Comment (luipir) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: syntax error closing the ) in the wrong place?

**Comment by luipir** [comment]
tnx @spatialthoughts for the core contribution :) may you add also a processing test for this alg? it's is especially important for core algs.

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 2 above)
Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: I started from VariableDistanceBuffer.py which had only Round translated. Fixed now

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 3 above)
Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: I don't see it. The script runs without error too. Can you specify where is the syntax error?

Comment (luipir) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]:
```suggestion
                                                         self.tr('Distance (in meters)'),
```
you ar right is not a syntax error is a simple typo

**Comment by spatialthoughts** [comment]
> tnx @spatialthoughts for the core contribution :) may you add also a processing test for this alg? it's is especially important for core algs.

Thanks for the quick review @luipir . I have added the test. This is the first time doing a processing test, so let me know if this looks ok.

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        self.miter_limit = None
+        self.join_styles = [self.tr('Round'),
+                            self.tr('Miter'),
+                            self.tr('Bevel')]
+        self.end_cap_styles = [self.tr('Round'),
+                       self.tr('Flat'),
+                       self.tr('Square')]
+
+    def initParameters(self, config=None):
+        self.addParameter(QgsProcessingParameterDistance(self.DISTANCE,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: You can call setDefaultUnit on this parameter with a unit type of meters to improve the ux

**Comment by luipir** [comment]
@spatialthoughts other than indentation seems you missed to add the alg in some cmake... AFAIK this would be the travis fail

**Comment by luipir** [comment]
and... tnx for the tests

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 4 above)
Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Nice! I was annoyed by the <unknown> label. This is much better. I have added it.

**Comment by spatialthoughts** [comment]
@luipir I have fixed the pep errors and now there is only 1 failing check. But I don't understand which cmake I missed. Any insights would be helpful.

**Comment by Gustry** [comment]
@luipir It's python, there is no cmake.

@spatialthoughts Your error is here https://travis-ci.org/qgis/QGIS/jobs/518718078#L2936

**Comment by luipir** [comment]
seems the registry can't find your alg... can't find why in this moment, it seems all correctly set

**Comment by m-kuhn** [comment]
add the alg to this list too https://github.com/qgis/QGIS/blob/8fa868a2e72a2af6b1b1ce70f1ac2cff65bae323/python/plugins/processing/algs/qgis/QgisAlgorithmProvider.py#L160

**Comment by spatialthoughts** [comment]
Thanks @m-kuhn . The test runs now but fails with a strange error. I don't know why it is trying to create a gpkg here

Processing(2): Could not create layer /tmp/tmploce9n5a/equidistance_buffer_points.gml'.gpkg: Creation of field fid failed (OGR error: Wrong field type for fid)
Processing(2): Could not create layer /tmp/tmploce9n5a/equidistance_buffer_points.gml'.gpkg: Creation of field fid failed (OGR error: Wrong field type for fid)

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        self.join_style = self.parameterAsEnum(
+            parameters, self.JOIN_STYLE, context) + 1
+        self.miter_limit = self.parameterAsDouble(
+            parameters, self.MITER_LIMIT, context)
+
+        source = self.parameterAsSource(parameters, 'INPUT', context)
+        self.source_crs = source.sourceCrs()
+        if not self.source_crs.isGeographic():
+            feedback.reportError(
+                'Layer CRS must be a Geograhpic CRS for this algorithm')
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Should this be translated?

Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Yes. Done.

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                       )
+
+from processing.algs.qgis.QgisAlgorithm import QgisFeatureBasedAlgorithm
+
+
+class EquidistanceBuffer(QgisFeatureBasedAlgorithm):
+    """
+    This algorithm creates equidistance buffers for vector layers.
+
+    Each geometry is transformed to a Azimuthal Equidistant projection centered
```
Comment (DelazJ) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Is it **a** Azimuthal or **an** Azimuthal?

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            parameters, self.JOIN_STYLE, context) + 1
+        self.miter_limit = self.parameterAsDouble(
+            parameters, self.MITER_LIMIT, context)
+
+        source = self.parameterAsSource(parameters, 'INPUT', context)
+        self.source_crs = source.sourceCrs()
+        if not self.source_crs.isGeographic():
+            feedback.reportError(
+                self.tr(
+                    'Layer CRS must be a Geograhpic CRS for this algorithm'))
```
Comment (DelazJ) [inline python/plugins/processing/algs/qgis/EquidistanceBuffer.py]:
```suggestion
                    'Layer CRS must be a Geographic CRS for this algorithm'))
```
and do we need the `for this algorithm` part?

Comment (m-kuhn) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: good point, it should be clear from the context that it concerns "this algorithm"

Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: fixed

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 6 above)
Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: I wrote it thinking 'a' refers to the projection and not azimuthal. A "Azimuthal equidistant' projection. I am not sure which is right though.

Comment (DelazJ) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: > I am not sure which is right though.

me neither ;)

**Comment by spatialthoughts** [comment]
I am still not sure what is causing the travis build to fail. The log has some entries which indicate that the test is trying to create geopackage and failing, but I don't know why the gml file would be converted to gpkg. Any suggestions?

**Comment by spatialthoughts** [comment]
@luipir @nyalldawson I would love to merge this, but not able to figure out how to fix the test. Any pointers?

File: python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml [hunk 8]
```diff
@@ -1386,6 +1386,22 @@ tests:
...
+      END_CAP_STYLE: 0
+      INPUT:
+        name: points.gml|layername=points
+        type: vector
+      JOIN_STYLE: 0
+      MITER_LIMIT: 2.0
+      SEGMENTS: 8
+    results:
+      OUTPUT:
+        name: expected/equidistance_buffer_points.gml'
```
Comment (luipir) [inline python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml]: as in https://travis-ci.org/qgis/QGIS/jobs/519176767#L1542 you have a ' more at the end of filename. Not clear to me is it's enough to fix the issue
```suggestion
        name: expected/equidistance_buffer_points.gml
```

**Comment by luipir** [comment]
hmmmm... here seems a testing environment error
https://travis-ci.org/qgis/QGIS/jobs/522770593#L1541
I suppose executing: 
https://github.com/qgis/QGIS/pull/9774/files#diff-7add1e33bfdf87577a677df3907a62f4R157
and from there there is the result mismatch.

I'm not sure I supposed well, btw I never found this kind of error. @nyalldawson @m-kuhn did you?

File: python/plugins/processing/algs/qgis/EquidistanceBuffer.py (hunk 6 above)
Comment (roya0045) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: An, if a vowel is after 'a ' it becomes 'an '.

Comment (spatialthoughts) [reply python/plugins/processing/algs/qgis/EquidistanceBuffer.py]: Thanks. @roya0045 and @DelazJ . I have replaced 'a' with 'an'.

**Comment by nyalldawson** [comment]
Can you please rebase this on master and I'll investigate the failing test?

**Comment by spatialthoughts** [comment]
@nyalldawson Thank you. I have rebased the branch to master. (at least I think I did). Hope you are able to investigate the error.

**Comment by nyalldawson** [comment]
Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

**Comment by luipir** [comment]
> Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

me too :(

**Comment by luipir** [comment]
> Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

btw now the error is completly changed... expected and result are different (starting from 4th or 5th digit).

**Comment by roya0045** [comment]
Have you tried with another crs? Would it be possible to do a rendering test instead of checking the exact coordinates?

Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.

**Comment by luipir** [comment]
> Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.

@roya0045  suppose is the way @spatialthoughts generated test data set... did you @spatialthoughts ?

**Comment by roya0045** [comment]
> > Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.
> 
> @roya0045 suppose is the way @spatialthoughts generated test data set... did you @spatialthoughts ?

Just to be more clear, what I meant by test is to use the travis test results instead of using 'home' results. This is what I do for my rendering tests since I want to make sure that nothing in my setup might generate data that can never be replicated in the branch and build tests, I just use the travis results and check if they are what I expected beforehand.

**Comment by spatialthoughts** [comment]
@roya0045 @luipir I generated the expected test data in QGIS running my algorithm. So I suppose there could be a mismatch between what is generated by Travis. How do I use the travis test results as the expected output? I can't figure out how to run this locally and grab the output.

**Comment by spatialthoughts** [comment]
I am still looking to merge this. Any suggestions on howto fix the test?

**Comment by luipir** [comment]
I do not have suggestions other thatn asking to @m-kuhn or @3nids . IMHO cloning this PR without tests in #30150 does not help. There is the rule to avoid merge if not covered by tests.

**Comment by luipir** [comment]
I can't reopen this PR because of #30150 of the same branch

**Comment by DelazJ** [comment]
And since we are in "feature freeze" period, this new alg can not be merged before release (3.10 is its target)

**Comment by luipir** [comment]
> And since we are in "feature freeze" period, this new alg can not be merged before release (3.10 is its target)

imho can be an exception... this is long term PR that have problem to merge (seems) due to testing infrastructure not depending on @spatialthoughts

**Comment by nyalldawson** [comment]
I have a concern about allowing this algorithm for non-point layers, because the results will be very misleading. The current approach just creates a CRS around the feature's centroid, so for large lines or polygon objects the results will NOT be accurate.

Given this, I don't feel comfortable allowing anything except single point layers to work with the algorithm -- otherwise we are just making it easy for users to silently create inaccurate results.

**Comment by luipir** [comment]
> Given this, I don't feel comfortable allowing anything except single point layers to work with the algorithm -- otherwise we are just making it easy for users to silently create inaccurate results.

@nyalldawson a layer with only a feature or a multi feature layer of a single part point geometry? 

@spatialthoughts can you rebase the code to fix conflicts?

**Comment by nyalldawson** [comment]
@luipir 
>a layer with only a feature or a multi feature layer of a single part point geometry?

No, just restricting to any point-based layers only (ie. no line or polygons).

But you raise an interesting point -- we also need to correctly consider multipoints here, by splitting them up, calculating the buffer, and the dissolving all the component buffers back to a single feature.

**Comment by spatialthoughts** [comment]
@nyalldawson I tested and geometry.buffer() works on multipoint geometry and does the correct thing by merging the buffers, so this algorithm doesn't need to do this separately. 

Restricting to point layers only sounds good.

I have got this branch in a state that is beyond my skill to rebase. I have a whole bunch of unrelated commits from past attempts and not sure how to proceed. Do you have any guidance? Starting fresh looks easier.

**Comment by nyalldawson** [comment]
@spatialthoughts 
> I tested and geometry.buffer() works on multipoint geometry and does the correct thing by merging the buffers, so this algorithm doesn't need to do this separately.

Right - but picture the situation with a multipoint feature, where one point sits on one side of a continent and another on the other side. In this case, it's not appropriate to create a single CRS for use when creating the buffer.

In this situation, you'd need to split the multipoint apart, and create the CRS and buffer for each point separately, and then finally dissolve the individual buffers back to a single multi-polygon object.

**Comment by nyalldawson** [comment]
I've tried to fix this branch too, without success. I'd suggest pulling out the files you changed and pasting them into a brand new branch based on master and abandoning this particular PR branch..

**Comment by spatialthoughts** [comment]
Closing this PR. Will submit a new PR from a new branch after fixing the issues mentioned earlier.

**Comment by ghtmtt** [comment]
@spatialthoughts I think this algorithm is worth to land QGIS. Do you have plans to resurrect it?

**Pull request closed without merging**